src/core/cpu_patches.cpp

```diff
@@ -0,0 +1,393 @@
+#else
+    const auto src = ZydisToXbyakMemoryOperand(operands[1]);
+
+    // Replace fs read with gs read.
+    c.putSeg(Xbyak::util::gs);
+    c.mov(dst, src);
+#endif
+}
+
+bool FilterAlwaysTrue(const ZydisDecodedOperand* operands) {
+    return true;
+}
+
+using PatchFilter = bool (*)(const ZydisDecodedOperand*);
+using InstructionGenerator = void (*)(const ZydisDecodedOperand*, Xbyak::CodeGenerator&);
+struct PatchInfo {
+    /// Filter for more granular patch conditions past just the instruction mnemonic.
+    PatchFilter filter;
+
+    /// Generator for the patch/trampoline.
+    InstructionGenerator generator;
+
+    /// Whether to use a trampoline for this patch.
+    bool trampoline;
+};
+
+static const std::unordered_map<ZydisMnemonic, PatchInfo> Patches = {
+#if defined(_WIN32) || defined(__APPLE__)
+    // Windows and Apple need a trampoline.
+    {ZYDIS_MNEMONIC_MOV, {FilterTcbAccess, GenerateTcbAccess, true}},
+#else
+    {ZYDIS_MNEMONIC_MOV, {FilterTcbAccess, GenerateTcbAccess, false}},
+#endif
+
+#ifdef __APPLE__
+    // BMI1 instructions that are not supported by Rosetta 2 on Apple Silicon.
+    {ZYDIS_MNEMONIC_ANDN, {FilterAlwaysTrue, GenerateANDN, true}},
+    {ZYDIS_MNEMONIC_BEXTR, {FilterAlwaysTrue, GenerateBEXTR, true}},
+    {ZYDIS_MNEMONIC_BLSI, {FilterAlwaysTrue, GenerateBLSI, true}},
+    {ZYDIS_MNEMONIC_BLSMSK, {FilterAlwaysTrue, GenerateBLSMSK, true}},
+    {ZYDIS_MNEMONIC_BLSR, {FilterAlwaysTrue, GenerateBLSR, true}},
+#endif
+};
+
+void PatchInstructions(u64 segment_addr, u64 segment_size, Xbyak::CodeGenerator& c) {
+    if (Patches.empty()) {
+        // Nothing to patch on this platform.
+        return;
+    }
+
+    ZydisDecoder instr_decoder;
+    ZydisDecodedInstruction instruction;
+    ZydisDecodedOperand operands[ZYDIS_MAX_OPERAND_COUNT];
+    ZydisDecoderInit(&instr_decoder, ZYDIS_MACHINE_MODE_LONG_64, ZYDIS_STACK_WIDTH_64);
+
+    u8* code = reinterpret_cast<u8*>(segment_addr);
+    u8* end = code + segment_size;
+    while (code < end) {
+        ZyanStatus status =
+            ZydisDecoderDecodeFull(&instr_decoder, code, end - code, &instruction, operands);
+        if (!ZYAN_SUCCESS(status)) {
+            code++;
+            continue;
+        }
+
+        if (Patches.contains(instruction.mnemonic)) {
+            auto patch_info = Patches.at(instruction.mnemonic);
+            if (patch_info.filter(operands)) {
+                auto patch_gen = Xbyak::CodeGenerator(instruction.length, code);
+
+                if (patch_info.trampoline) {
+                    const auto trampoline_ptr = c.getCurr();
+
+                    patch_info.generator(operands, c);
+
+                    // Return to the following instruction at the end of the trampoline.
+                    c.jmp(code + instruction.length);
+
+                    // Replace instruction with near jump to the trampoline.
+                    patch_gen.jmp(trampoline_ptr, Xbyak::CodeGenerator::LabelType::T_NEAR);
```

Comment:
I don't see the prefix bytes to nop patch anymore. This will cause some openorbis demos to crash if you leave it out

Reply:
The nop bytes are done further down, it's set up so that you either patch in a new instruction or a jump to a trampoline, then regardless of which happens it will fill the remainder with nops.

Reply:
I see, I think zydis will handle it properly in this case as it fully parses the instructions

Reply:
Ah sorry I misread, yeah the prefix bytes should be handled by Zydis, and then any remaining data gets nop'd out.